Overlay Regions should be updated when loading a PDF

etiennesegonzac opened the pull request:
#### 16b1f218b34090cb7e7f23569bebd49966b7df45
<pre>
Overlay Regions should be updated when loading a PDF
<a href="https://bugs.webkit.org/show_bug.cgi?id=277556">https://bugs.webkit.org/show_bug.cgi?id=277556</a>
&lt;<a href="https://rdar.apple.com/125667363">rdar://125667363</a>&gt;

Reviewed by Abrar Rahman Protyasha.

We need to observe the contentSize updates made from PDFKit to properly
maintain Overlay Regions when a PDF is loaded.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateOverlayRegionsForCustomContentView]):
Introduce a new method to refresh the Overlay Region settings for custom
content views.

* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView web_initWithFrame:webView:mimeType:]):
Add an observer on the scrollview&apos;s contentSize.
(-[WKPDFView dealloc]):
Remove the observer.
(-[WKPDFView observeValueForKeyPath:ofObject:change:context:]):
Update Overlay Regions when the contentSize is updated.

* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(-[WKBaseScrollView _updateOverlayRegionRects:]):
Switch to a const ref.

Canonical link: <a href="https://commits.webkit.org/281936@main">https://commits.webkit.org/281936@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78816f060365b911115fe221050820e4f67b130e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61376 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40736 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13960 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65328 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11924 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48414 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12199 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49580 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8280 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64204 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37869 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53162 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30423 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34530 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10837 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56338 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10690 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67059 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5322 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10464 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56954 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5345 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53127 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57169 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13700 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4394 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36540 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37623 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38717 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37367 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1323,6 +1323,18 @@ - (void)_updateOverlayRegions:(const WebKit::LayerPropertiesMap&)changedLayerPro
+- (void)_updateOverlayRegionsForCustomContentView
+{
+    if (![self _scrollViewCanHaveOverlayRegions:_scrollView.get()]) {
+        [_scrollView _updateOverlayRegionsBehavior:false];
```
Should use `NO` instead of `false`, since the method takes a `BOOL` parameter.

I think there are some existing cases that could be fixed as well.

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1323,6 +1323,18 @@ - (void)_updateOverlayRegions:(const WebKit::LayerPropertiesMap&)changedLayerPro
+        return;
+    }
+
+    [_scrollView _updateOverlayRegionsBehavior:true];
```
Ditto.

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1323,6 +1323,18 @@ - (void)_updateOverlayRegions:(const WebKit::LayerPropertiesMap&)changedLayerPro
+
+    [_scrollView _updateOverlayRegionsBehavior:true];
+    HashSet<WebCore::IntRect> empty;
+    [_scrollView.get() _updateOverlayRegionRects:empty];
```
`.get()` is unnecessary.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKPDFView.mm:
```diff
@@ -199,6 +208,10 @@ - (instancetype)web_initWithFrame:(CGRect)frame webView:(WKWebView *)webView mim
     _webView = webView;
 
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+    [[_webView _wkScrollView] addObserver:self forKeyPath:@"contentSize" options:NSKeyValueObservingOptionNew context:kvoContext];
```
I wonder if it's sufficient to override `setContentSize:` rather than use KVO?

aprotyas replied:
(same question)

aprotyas commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1323,6 +1323,18 @@ - (void)_updateOverlayRegions:(const WebKit::LayerPropertiesMap&)changedLayerPro
+
+    [_scrollView _updateOverlayRegionsBehavior:YES];
+    HashSet<WebCore::IntRect> empty;
+    [_scrollView _updateOverlayRegionRects:empty];
```
Random remark out of the scope of this PR, but tangentially related; could we change `-[WKBaseScrollView _updateOverlayRegionRects:]` to accept a const ref instead? Then you could pass in a temporary and make the call a little more succinct?

```cpp
@interface WKBaseScrollView : WKBEScrollView
- (void)_updateOverlayRegionRects:(const HashSet<WebCore::IntRect>&)overlayRegions;
@end

// ...
[_scrollView _updateOverlayRegionRects:{ }];
```

etiennesegonzac replied:
that's a nice clean-up! updating the patch

etiennesegonzac committed:
Overlay Regions should be updated when loading a PDF
https://bugs.webkit.org/show_bug.cgi?id=277556
<rdar://125667363>

Reviewed by Abrar Rahman Protyasha.

We need to observe the contentSize updates made from PDFKit to properly
maintain Overlay Regions when a PDF is loaded.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateOverlayRegionsForCustomContentView]):
Introduce a new method to refresh the Overlay Region settings for custom
content views.

* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView web_initWithFrame:webView:mimeType:]):
Add an observer on the scrollview's contentSize.
(-[WKPDFView dealloc]):
Remove the observer.
(-[WKPDFView observeValueForKeyPath:ofObject:change:context:]):
Update Overlay Regions when the contentSize is updated.

* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(-[WKBaseScrollView _updateOverlayRegionRects:]):
Switch to a const ref.

Canonical link: https://commits.webkit.org/281936@main
